Enable Compatibility Layer for Product Collection

kmanijak opened the pull request:
Product Collection has to be compatible with the extensions and cover the same extensibility points as Products (Beta) block does.

In this PR Compatibility Layer is enabled for Product Collection block.

Fixes https://github.com/woocommerce/woocommerce-blocks/issues/10194 and https://github.com/woocommerce/woocommerce-blocks/issues/10378

### Testing

#### Automated Tests
* [ ] Changes in this PR are covered by Automated Tests.
  * [ ] Unit tests
  * [ ] E2E tests

#### User Facing Testing

<!-- Write these steps from the perspective of a "user" (merchant) familiar with WooCommerce. No need to spell out the steps for common setup scenarios (eg. "Create a product"), but do be specific about the thing being tested. Include screenshots demonstrating expectations where that will be helpful. -->

##### Core tests - include in the testing steps (fix for https://github.com/woocommerce/woocommerce-blocks/issues/10378)

1. Go to Editor
2. Add Products (Beta) block
3. Add the following code to your site using the [Code Snippets](https://wordpress.org/plugins/code-snippets/) plugin:
```
add_action( 'woocommerce_before_shop_loop_item' , function () {
	echo 'Hook: woocommerce_before_shop_loop_item';
});

add_action( 'woocommerce_after_shop_loop_item' , function () {
	echo 'Hook: woocommerce_after_shop_loop_item';
});
```

4. Go to the frontend

**Expected:** The content is rendered above and below each product

##### Experimental tests - do not include in testing steps (fix for https://github.com/woocommerce/woocommerce-blocks/issues/10194)

1. Go to Editor and edit Product Catalog and Product Search Results 
2. Clear customizations in both of them
4. If the Classic Template shows up, transform it to blockified template
5. Replace Products (Beta) block with Product Collection
6. Add Product Search block to the Header
7. Save templates
8. Add the following code to your site using the [Code Snippets](https://wordpress.org/plugins/code-snippets/) plugin, then reload the shop page on the front end to see if the test content is displayed at the expected position:
```
add_action( 'woocommerce_before_shop_loop_item' , function () {
	echo 'Hook: woocommerce_before_shop_loop_item';
});
```

Repeat for other hooks (script covering all of those actions at the bottom of PR for better readibility):
- [ ] `woocommerce_before_main_content`
- [ ] `woocommerce_after_main_content`
- [ ] `woocommerce_before_shop_loop_item_title`
- [ ] `woocommerce_shop_loop_item_title`
- [ ] `woocommerce_after_shop_loop_item_title`
- [ ] `woocommerce_before_shop_loop_item`
- [ ] `woocommerce_after_shop_loop_item`
- [ ] `woocommerce_before_shop_loop`
- [ ] `woocommerce_after_shop_loop`
- [ ] `woocommerce_no_products_found` - to check this scenario, you can use Product Search and type some phrase that won't return any results (e.g. "asdfasdf").

Double check these still work for Products (Beta) block by removing Product Collection and adding Products (Beta) block back.

* [x] Do not include in the Testing Notes <!-- Check this box if this PR can't be tested by users (ie: it doesn't include user-facing changes or it can't be tested without manually modifying the code). -->

### WooCommerce Visibility

<!-- Check this [this doc](../docs/blocks/feature-flags-and-experimental-interfaces.md) to see if the change is visible in WC core or not (part of the feature plugin or experimental)-->

* [x] WooCommerce Core
* [ ] Feature plugin
* [x] Experimental

---

Appendix - script for testing steps

```
add_action( 'woocommerce_before_main_content' , function () {
	echo 'Hook: woocommerce_before_main_content';
});

add_action( 'woocommerce_after_main_content' , function () {
	echo 'Hook: woocommerce_after_main_content';
});

add_action( 'woocommerce_before_shop_loop_item_title' , function () {
	echo 'Hook: woocommerce_before_shop_loop_item_title';
});

add_action( 'woocommerce_shop_loop_item_title' , function () {
	echo 'Hook: woocommerce_shop_loop_item_title';
});

add_action( 'woocommerce_after_shop_loop_item_title' , function () {
	echo 'Hook: woocommerce_after_shop_loop_item_title';
});

add_action( 'woocommerce_before_shop_loop_item' , function () {
	echo 'Hook: woocommerce_before_shop_loop_item';
});

add_action( 'woocommerce_after_shop_loop_item' , function () {
	echo 'Hook: woocommerce_after_shop_loop_item';
});

add_action( 'woocommerce_before_shop_loop' , function () {
	echo 'Hook: woocommerce_before_shop_loop';
});

add_action( 'woocommerce_after_shop_loop' , function () {
	echo 'Hook: woocommerce_after_shop_loop';
});

add_action( 'woocommerce_no_products_found' , function () {
	echo 'Hook: woocommerce_no_products_found';
});
```

### Changelog

> Products: fix the before/after shop item hooks

kmanijak committed:
Enable Compatibility Layer for Product Collection

kmanijak committed:
Merge branch 'trunk' into add/compatibility-layer-to-product-collection

kmanijak commented:
Perhaps better testing steps would be to trigger hooks programmatically rather than by using extensions 🤔 What do you think @gigitux?

gigitux commented:
> Perhaps better testing steps would be to trigger hooks programmatically rather than by using extensions 🤔 What do you think @gigitux?

I suggest following the Tung approach #8172. Furthermore, it would be great to add some E2E tests. In the next few days, in the trunk it will be available an easy way to test hook with playwright: https://github.com/woocommerce/woocommerce-blocks/pull/10206

kmanijak committed:
Add another condition to recognise shop item

kmanijak commented:
> I suggest following the Tung approach https://github.com/woocommerce/woocommerce-blocks/pull/8172. Furthermore, it would be great to add some E2E tests. In the next few days, in the trunk it will be available an easy way to test hook with playwright: https://github.com/woocommerce/woocommerce-blocks/pull/10206

Thanks for this suggestion @gigitux! It was really helpful as I actually found two issues:

1. Adding hooks for Product Collection disables some hooks for Products (Beta) block
2. `woocommerce_before_shop_loop_item` and `woocommerce_after_shop_loop_item` do not work either for Product Collection or Products (Beta) on `trunk`

I'll raise an issue for 2., but 1. has to be fixed within this PR, so I'm switching this PR to draft until I fix it.

kmanijak committed:
Enable multiple blocks to be assigned to single hook in Compatibility Layer

kmanijak committed:
Move core/null block name assignment before the hooks assignment which is based on the block name

kmanijak commented:
@gigitux I fixed both issues:

> 1. Adding hooks for Product Collection disables some hooks for Products (Beta) block

Commit: [here](https://github.com/woocommerce/woocommerce-blocks/pull/10332/commits/34dcf3f64216315d77c70d6443c5cda7521e868c)
Solution: Previously I added new items to array with the same keys which simply overridden the previous keys. In this commit I change the logic a bit, so in the hooks array you can define a set of blocks to which the hook should be applied.

> 2. `woocommerce_before_shop_loop_item` and `woocommerce_after_shop_loop_item` do not work either for Product Collection or Products (Beta) on `trunk`

Commit: [here](https://github.com/woocommerce/woocommerce-blocks/pull/10332/commits/c038a042af7c5944f1a74642ee1e699ab48c1e6f)
Solution: There's logic related to abstract `core/null` block that represents the single product. It has "artificial" block_name assigned and assignment happened AFTER the hooks where grouped based on block_name.

> Furthermore, it would be great to add some E2E tests. In the next few days, in the trunk it will be available an easy way to test hook with playwright: https://github.com/woocommerce/woocommerce-blocks/pull/10206

Regarding tests I'm planning to add them next week (Cooldown), but this documentation looks promising!

kmanijak committed:
Merge branch 'trunk' into add/compatibility-layer-to-product-collection

gigitux commented on src/Templates/ArchiveProductTemplatesCompatibility.php:
```diff
@@ -93,24 +97,13 @@ function( $hook ) use ( $block_name ) {
-			},
-			array_values( $this->hook_data )
+		$supported_blocks = array_merge(
+			...array_map(
```
The spread operator is available on PHP 7.4 ([source](https://stackoverflow.com/a/56072765)). Currently, the minium version of PHP supported by WooCommerce is PHP 7.3 https://developer.woocommerce.com/2023/06/05/new-requirement-for-woocommerce-8-2-php-7-4/

I think that we should revisit this line.

kmanijak replied:
AFAIK in this context it will work, as spreading array as an argument to another function was introduced in PHP 5.6 ([StackOverflow](https://stackoverflow.com/a/46861938), [PHP docs](https://www.php.net/manual/en/migration56.new-features.php) - look for "splat operator").

What was introduced in PHP 7.4 was a spread operator in array expressions ([docs](https://wiki.php.net/rfc/spread_operator_for_array)) that allows spreading one array into another one, like:

```
$arr = [3, 4];
$numbers = [1, 2, ...$arr, 5, 6]; // -> [1, 2, 3, 4, 5, 6]
```

From the docs:
> PHP has already supported [argument unpacking](https://wiki.php.net/rfc/argument_unpacking) (AKA spread operator) since 5.6. This RFC proposes to bring this feature to array expression.

---

However, I'll add an empty array as an argument to `array_merge` before spreading another array to make it safer in case we spread an empty array.

gigitux replied:
TIL! Thanks for making the code safer 👍

gigitux commented on src/Templates/ArchiveProductTemplatesCompatibility.php:
```diff
@@ -182,60 +175,60 @@ function( $hook ) {
-				'block_name' => 'core/query',
-				'position'   => 'before',
-				'hooked'     => array(
+				'block_names' => array( 'core/query', 'woocommerce/product-collection' ),
```
This is my personal preference, but I would keep the same data structure across the compatibility layers. So, I would update the SingleProductTemplateCompatibility file as well.

kmanijak replied:
I agree, addressed in [this commit](https://github.com/woocommerce/woocommerce-blocks/pull/10332/commits/bfadbd47ffefd2afed69718e6a7298fa60721b23).

gigitux replied:
Could you update [this comment](https://github.com/woocommerce/woocommerce-blocks/blob/6339f97e70b207a677f2f568c9b749c790eccc4c/src/Templates/AbstractTemplateCompatibility.php/#L99-L121) in the AbstractTemplateCompatibility?

gigitux reviewed (changes requested):
Thanks for your great work! 🚀 

I added two comments. One should be addressed. For the second one, I want to get your feedback about it.

Also, I noticed that with your changes, the `woocommerce_before_shop_loop` and `woocommerce_after_shop_loop` hooks are injected into the page even if any product hasn't been found. For the Products block, I fixed this in a dedicated PR: https://github.com/woocommerce/woocommerce-blocks/pull/9464. I think that the approach should be the same.

kmanijak committed:
Make array operations safer

kmanijak committed:
Unify hooks interface between templates

kmanijak committed:
Merge branch 'trunk' into add/compatibility-layer-to-product-collection

kmanijak committed:
Fix typo

kmanijak commented:
> Also, I noticed that with your changes, the woocommerce_before_shop_loop and woocommerce_after_shop_loop hooks are injected into the page even if any product hasn't been found. For the Products block, I fixed this in a dedicated PR: https://github.com/woocommerce/woocommerce-blocks/pull/9464. I think that the approach should be the same.

@gigitux, I didn't change that logic, though, this check is still in [here](https://github.com/woocommerce/woocommerce-blocks/pull/10332/files#diff-67e968b526ce5d483df5b3830c9e2519b0adff6a3801648efa025d23d733d2b6R88).

I double-checked in the `trunk` and here's the outcome:

Products (`trunk`) | Products (this branch) | Product Collection
-- | -- | --
<img width="1464" alt="image" src="https://github.com/woocommerce/woocommerce-blocks/assets/20098064/48a81ede-0a31-4c8d-8757-c215f9944367"> | <img width="1465" alt="image" src="https://github.com/woocommerce/woocommerce-blocks/assets/20098064/17eeb682-034e-4473-a0d1-3879b71c038c"> | <img width="1464" alt="image" src="https://github.com/woocommerce/woocommerce-blocks/assets/20098064/56e1b028-19d6-48ec-9d04-f610ff0de3d4">


As you can see, the same hooked content appears on the page in each case. If you're saying that `woocommerce_before_shop_loop_item` and `woocommerce_after_shop_loop_item` should not be rendered in this case, I'd raise a separate issue to not block this PR. Unless I should test it differently?

kmanijak commented:
I created an issue for the problem described in [this comment](https://github.com/woocommerce/woocommerce-blocks/pull/10332#issuecomment-1659029151): https://github.com/woocommerce/woocommerce-blocks/issues/10452

EDIT: and the fix: https://github.com/woocommerce/woocommerce-blocks/pull/10453

gigitux reviewed (approved):
Great work! LGTM! I just added a new comment about updating the comments on the AbstractTemplateCompatibility class (https://github.com/woocommerce/woocommerce-blocks/blob/6339f97e70b207a677f2f568c9b749c790eccc4c/src/Templates/AbstractTemplateCompatibility.php/#L99-L121).

BTW, I will pre-approve this PR! 🚀

kmanijak committed:
Update comment describing set_hook_data to reflect new interface

kmanijak committed:
Merge branch 'trunk' into add/compatibility-layer-to-product-collection
